torch/_inductor/decomposition.py

```diff
@@ -741,3 +741,76 @@ def max_pool2d_with_indices(
         padding,
     )
     return vals, indices
+
+
+def _use_online_softmax(x, dim):
+    if not config.online_softmax:
+        return False
+
+    # Don't do online softmax for scalar or 1d tensor
+    if x.dim() < 2:
+        return False
+
+    # Only do online softmax for GPU for now
+    if x.device.type != "cuda":
+        return False
+
+    return x.size(dim) > 2**14
```

Comment:
This is going to add a guard.. is that we want ?

It's also going to cause all other pattern matching that involves softmax to fail after x.size(dim) > 2 ** 14 to fail, such as SDPA.

Can we instead do this as a pattern match that runs after other passes ?

Reply:
> This is going to add a guard.. is that we want ?

Good point. Any suggestion to improve it?

Reply:
> It's also going to cause all other pattern matching that involves softmax to fail after x.size(dim) > 2 ** 14 to fail, such as SDPA.

I give this another thought. But even if the online softmax decomposition rule kicks in for SDPA, this is still fine? Because pattern should also generated Fx graphs with online softmax. This does not affect if the pattern can get a match or not. Wdyt?

I can also add some tests for this.

Reply:
hmm, the example inputs to trace the pattern is quite small (not the same as the inputs the the softmax in the model will get). That may cause some diecrepancy.

Reply:
How does the performance compare on smaller shapes?

Reply:
> How does the performance compare on smaller shapes? 

May I know why we care about perf on smaller shapes?

The 2 major usages of softmax I care most are:
1. those used in the attention layer, which we should leverage fused attention like flash attention
2. the softmax inside CrossEntryLoss. In which case, the reduction dimension is the vocab size which should most likely be larger than the 16K threshold here.

Do I miss some other important cases?

Reply:
If it works well on smaller cases then we can just always turn it on.

Reply:
> I give this another thought. But even if the online softmax decomposition rule kicks in for SDPA, this is still fine? Because the pattern should also generated Fx graphs with online softmax. This does not affect if the pattern can get a match or not. Wdyt?

The difficulty is that it only sometimes fires.

Reply:
Do we need to enable persistent kernel/split reduction to always turn it on ? 

Also, doing a late pattern is not too difficult. it might be an easier first step to landing

Reply:
Re: size. you could just disable if there are symints. or there is a suppress_guards api